src/types/mutex.rs

```diff
@@ -0,0 +1,375 @@
+/// thread is blocked on acquiring the lock. On the free-threaded build, threads
+/// blocked on acquiring a PyMutex will not prevent the garbage collector from
+/// running.
+///
+/// ## Poisoning
+///
+/// Like `std::sync::Mutex`, `PyMutex` implements poisoning. A mutex
+/// is considered poisoned whenever a thread panics while holding the mtex. Once
+/// a mutex is poisoned, all other threads are unable to access the data by
+/// default as it is likely to be tainted (some invariant is not being held).
+///
+/// This means that the `lock` method returns a `Result` which indicated whether
+/// the mutex has been poisoned or not. Must usage will simple `unwrap()` these
+/// results, propagating panics among threads to ensure a possible invalid
+/// invariant is not being observed.
+///
+/// A poisoned mutex, however, does not prevent all access to the underlying
+/// data. The `PoisonError` type has an `into_inner` method which will return
+/// the guard that would have otherwise been returned on a successful lock. This
+/// allows access to the data, despite the lock being poisoned.
+pub struct PyMutex<T: ?Sized> {
+    pub(crate) mutex: UnsafeCell<crate::ffi::PyMutex>,
+    poison: Flag,
+    pub(crate) data: UnsafeCell<T>,
+}
+
+/// RAII guard to handle releasing a PyMutex lock.
+///
+/// The lock is released when `PyMutexGuard` is dropped.
+pub struct PyMutexGuard<'a, T: ?Sized> {
+    inner: &'a PyMutex<T>,
+    poison: Guard,
+}
+
+/// `T` must be `Sync` for a [`PyMutexGuard<T>`] to be `Sync`
+/// because it is possible to get a `&T` from `&MutexGuard` (via `Deref`).
+unsafe impl<T: ?Sized + Sync> Sync for PyMutexGuard<'_, T> {}
+
+/// `T` must be `Send` for a [`Mutex`] to be `Send` because it is possible to acquire
+/// the owned `T` from the `Mutex` via [`into_inner`].
+///
+/// [`into_inner`]: Mutex::into_inner
+unsafe impl<T: ?Sized + Send> Send for PyMutex<T> {}
+
+/// `T` must be `Send` for [`PyMutex`] to be `Sync`.
+/// This ensures that the protected data can be accessed safely from multiple threads
+/// without causing data races or other unsafe behavior.
+///
+/// [`PyMutex<T>`] provides mutable access to `T` to one thread at a time. However, it's essential
+/// for `T` to be `Send` because it's not safe for non-`Send` structures to be accessed in
+/// this manner. For instance, consider [`Rc`], a non-atomic reference counted smart pointer,
+/// which is not `Send`. With `Rc`, we can have multiple copies pointing to the same heap
+/// allocation with a non-atomic reference count. If we were to use `Mutex<Rc<_>>`, it would
+/// only protect one instance of `Rc` from shared access, leaving other copies vulnerable
+/// to potential data races.
+///
+/// Also note that it is not necessary for `T` to be `Sync` as `&T` is only made available
+/// to one thread at a time if `T` is not `Sync`.
+///
+/// [`Rc`]: crate::rc::Rc
+unsafe impl<T: ?Sized + Send> Sync for PyMutex<T> {}
```

Comment:
all these unsafe impls were copied from `std::sync::mutex`. I didn't copy [this one](https://github.com/rust-lang/rust/blob/246733a3d978de41c5b77b8120ba8f41592df9f1/library/std/src/sync/poison/mutex.rs#L186) because I don't think `PyMutex` has this restriction? This is all black magic to me so I'd very much appreciate input on what's correct here.